crates/sui-core/src/quorum_driver/reconfig_observer.rs

```diff
@@ -0,0 +1,104 @@
+    reconfig_rx: tokio::sync::broadcast::Receiver<Committee>,
+    authority_store: Arc<AuthorityStore>,
+    committee_store: Arc<CommitteeStore>,
+    safe_client_metrics_base: SafeClientMetricsBase,
+    auth_agg_metrics: AuthAggMetrics,
+}
+
+impl OnsiteReconfigObserver {
+    pub fn new(
+        reconfig_rx: tokio::sync::broadcast::Receiver<Committee>,
+        authority_store: Arc<AuthorityStore>,
+        committee_store: Arc<CommitteeStore>,
+        safe_client_metrics_base: SafeClientMetricsBase,
+        auth_agg_metrics: AuthAggMetrics,
+    ) -> Self {
+        Self {
+            reconfig_rx,
+            authority_store,
+            committee_store,
+            safe_client_metrics_base,
+            auth_agg_metrics,
+        }
+    }
+
+    async fn create_authority_aggregator_from_system_state(
+        &self,
+    ) -> AuthorityAggregator<NetworkAuthorityClient> {
+        AuthorityAggregator::new_from_system_state(
+            &self.authority_store,
+            &self.committee_store,
+            self.safe_client_metrics_base.clone(),
+            self.auth_agg_metrics.clone(),
+        )
+        // TODO: we should tolerate when <= f validators give invalid addresses
+        // GH issue: https://github.com/MystenLabs/sui/issues/7019
+        .unwrap_or_else(|e| {
+            panic!(
+                "Failed to create AuthorityAggregator from System State: {:?}",
+                e
+            )
+        })
+    }
+}
+
+#[async_trait]
+impl ReconfigObserver<NetworkAuthorityClient> for OnsiteReconfigObserver {
+    async fn run(&mut self, quorum_driver: Arc<QuorumDriver<NetworkAuthorityClient>>) {
+        // A tiny optimization: when a very stale node just starts, the
+        // channel may fill up committees quickly. Here we skip directly to
+        // the last known committee by looking at SuiSystemState.
+        let authority_agg = self.create_authority_aggregator_from_system_state().await;
+        if authority_agg.committee.epoch > quorum_driver.current_epoch() {
+            quorum_driver
+                .update_validators(Arc::new(authority_agg))
+                .await;
+        }
+        loop {
+            match self.reconfig_rx.recv().await {
+                Ok(committee) => {
+                    info!("Got reconfig message: {}", committee);
+                    if committee.epoch > quorum_driver.current_epoch() {
+                        let authority_agg =
+                            self.create_authority_aggregator_from_system_state().await;
+                        quorum_driver
+                            .update_validators(Arc::new(authority_agg))
+                            .await;
+                    } else {
+                        // This should only happen when the node just starts
+                        warn!("Ignored non-newer from reconfig channel: {}", committee);
+                    }
+                }
+                // Neither closed channel nor lagged shall happen
```

Comment:
why not just handle the lagged state? it should be recoverable. its not clear to me why it can't possibly happen.

Reply:
Similarly for closed channel. Is there a reason for this assumption? Probably best to handle it in case we change the way executor works. As an example, we were toying with having executor restart at epoch end to make strong consistency of epoch store easier. This would create problems for you if you assume that this channel cannot close from underneath you